arraykit.c

```diff
@@ -129,6 +130,62 @@ AK_ResolveDTypeIter(PyObject *dtypes)
     return resolved;
 }
 
+// Numpy implementation: https://github.com/numpy/numpy/blob/a14c41264855e44ebd6187d7541b5b8d59bb32cb/numpy/core/src/multiarray/methods.c#L1557
+PyObject*
+AK_ArrayCopy(PyObject *array, PyObject *memo)
+{
+    PyObject *id = PyLong_FromVoidPtr(array);
+    PyObject *found = PyDict_GetItem(memo, id);
+    if (found) { // found will be NULL if not in dict
+        Py_DECREF(id);
+        Py_INCREF(found); // got a borrowed ref
+        return found;
+    }
+
+    // if dtype is object, call deepcopy with memo
+    PyObject *array_new;
+    PyArray_Descr *dtype = PyArray_DESCR((PyArrayObject *)array); // borrowed ref
+
+    if PyDataType_ISOBJECT(dtype)
+    {
```

Comment:
This currently only works because the macro definition is parenthesized!
```suggestion
    if (PyDataType_ISOBJECT(dtype)) {
```

Reply:
Nice catch! Python slipping into my C...